HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,49 @@
+# import findspark
+# findspark.init()
+import pyspark
+import numpy as np
+
+
+def copartitioned(RDD1, RDD2):
+    "check if two RDDs are copartitioned"
+    return RDD1.partitioner == RDD2.partitioner
+
+
+def filterer(i):
+    def filt(x):
+        return x==itr_acc.value
+    return filt
+
+def get_first_non_empty_cogroup(x2):
+    for elem in x2:
+        elem = list(elem)
+        if len(elem) > 0:
+            return elem[0]
+
+
+#Breadth-first search. N corresponds to the number of partitions
+def BFS(graph, starting_node, N):
+    
+    nodes = sc.parallelize([(starting_node, 1)]).partitionBy(N)
+    
+    graph.partitionBy(N).cache()
+    
+    total_char_acc = sc.accumulator(1)
+    it_acc = sc.accumulator(0)
+
+    while True:
+        it_acc += 1
+        dist = it_acc.value
+
+        nodes_2 = nodes.filter(lambda x: filterer(x[1])).join(graph).flatMap(lambda x: x[1][1]).distinct().map(lambda x: (x, dist + 1)).partitionBy(N)
+
+        nodes = nodes.cogroup(nodes_2).map(lambda x: (x[0], get_first_non_empty_cogroup(list(x[1])))).partitionBy(N).cache()
+        
+        assert copartitioned(nodes_2, nodes)
+
+        if nodes.count() == total_char_acc.value or dist == 10:
+            break
```

Comment:
^ Not sure I follow your point.  This statement is in the while loop, and you quit here since the loop's condition is always True and will never end on its own